### ISSUE TITLE
A faster way to list sheets in a workbook?

### ISSUE THREAD

**Issue by perlancar** [Tux/Spreadsheet-Read#49]
Hi,

Thanks for this wonderful module. I'm using Spreadsheet::Read to list sheets using something like:

    my $wb = Spreadsheet::Read->new($filename);
    my $sheets = [$wb->sheets];

for shell tab completion purposes [1], e.g. `ss2csv foo.xlsx -s <tab>`. Unfortunately, for non-trivial files it's rather slow. For example, I have an .xlsx file with size 181kb, 3 sheets, total ~1700 rows in all sheets and it takes ~4 seconds on my laptop for the completion to show up.

I'm wondering if there is a faster way to list sheets in a workbook?

[1] https://metacpan.org/dist/App-SpreadsheetUtils/view/script/ss2csv

**Comment by Tux** [comment]
See the very first item in [https://github.com/Tux/Spreadsheet-Read/blob/master/doc/Spreadsheet-Read.md#todo|TODO].

As Spreadsheet::Read is 100% depending on the underlying parsers, these should **also** support that and they don't. e.g. An XLSX or ODS is a ZIP file with embedded XML, and that all has to be read in order to be able to create the index. This is one of the disadvantages of the new(er) formats. In XLS this was easy.

In other words: patches welcome, but the patches would be to the parsers first in order to have this wrapper be able to use it

**Comment by Tux** [comment]
Closed as you now can use `ReadData ($file, parser => "Excel::ValueReader::XLSX");` if you really just need the index

**Comment by perlancar** [comment]
Thanks. It's not fast enough to make tab completion less annoying, but at least it's more bearable. Some benchmarks (the sample XLSX files are on CPAN in Spreadsheet-Examples):

09Nov15:39:23 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "empty.xlsx"), parser => "Excel::ValueReader::XLSX")' )

real    0m0.401s
user    0m0.360s
sys     0m0.040s
09Nov15:39:31 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-empty.xlsx"), parser => "Excel::ValueReader::XLSX")' )

real    0m0.408s
user    0m0.349s
sys     0m0.034s
09Nov15:39:56 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-10row-10col.xlsx"), parser => "Excel::ValueReader::XLSX")' )

real    0m0.375s
user    0m0.334s
sys     0m0.040s
09Nov15:40:16 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-1000row-10col.xlsx"), parser => "Excel::ValueReader::XLSX")' )

real    0m1.297s
user    0m1.245s
sys     0m0.050s

Compared to the default parser:

09Nov15:40:22 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "empty.xlsx"))' )
real    0m0.224s
user    0m0.218s
sys     0m0.006s
09Nov15:41:44 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-empty.xlsx"))' )

real    0m0.296s
user    0m0.281s
sys     0m0.014s
09Nov15:42:01 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-10row-10col.xlsx"))' )

real    0m0.422s
user    0m0.379s
sys     0m0.044s
09Nov15:42:10 u1@u1-XPS-13-9360:~$ time ( perl -MDD -MSpreadsheet::Read -MFile::ShareDir=:ALL -E'$res = ReadData (dist_file("Spreadsheet-Examples", "10sheet-1000row-10col.xlsx"))' )

real    0m15.742s
user    0m15.590s
sys     0m0.138s

I think I can create a stripped down version of Excel::ValueReader::XLSX (without Moose, Date::Calc, etc) just to list sheets.

**Comment by Tux** [comment]
I'm willing to implement an interface to any sane backend that matches a minimum API to Spreadsheet::Read **and does not add a truckload of dependencies**.
I admire your motivation to drop your work on CPAN, but I am unwilling to install all your modules to test some basic functionality. Likewise for authors that break "normal" behavior, like Marc Lehman with their modules all require the unacceptable "common::sense" (which has no common sense whatsoever)

**Comment by perlancar** [comment]
> I'm willing to implement an interface to any sane backend that matches a minimum API to Spreadsheet::Read **and does not add a truckload of dependencies**. I admire your motivation to drop your work on CPAN, but I am unwilling to install all your modules to test some basic functionality. Likewise for authors that break "normal" behavior, like Marc Lehman with their modules all require the unacceptable "common::sense" (which has no common sense whatsoever)

Thanks for the heads up, though I think it's a bit loaded with assumptions. 

First, I don't add dependencies willy nilly. If you look at some of my modules, e.g. CLI apps, they do use my framework Perinci::CmdLine (which has a lot of dependencies currently, unfortunately) to speed up development. But I do create minimal and light modules with little to no dependencies e.g. Log::ger, etc as needed.

Second, if I created a stripped-down version of  Excel::ValueReader::XLSX to just list sheets or extract other information as needed for the purpose of tab completion, it probably wouldn't be suitable for a backend for Spreadsheet::Read anyway.
